Add documentation for EIB's RPM resolution logic (#161)
suse-edge/edge-image-builder: docs/installing-packages.md

# Installing packages
This documentation dives deeper into how a user can configure packages for installation inside the EIB image. Furthermore, it explains the **RPM resolution** process that EIB goes through so it can ensure that configured packages will be successfully installed even in an **air-gapped** environment.

> Below you can find a table of contents for easier navigation:

Reviewer: This is arguably not needed. Go to https://github.com/suse-edge/edge-image-builder/blob/main/docs/building-images.md and click the button on the right that has three rows of a dot and a line. GitHub automatically generates a TOC based on the headings.

1. [Supported systems](#supported-systems)

> 2. [Configure packages for installation](#configure-packages-for-installation)

Reviewer: Does it work to use the built-in syntax of prefixing all items in a numbered list as `1.`? I'm not sure if the embedded unordered list messes that up and is why you manually number them. If we can, it's better to use the markdown syntax where possible so we can take advantage of any accessibility features it provides.

    * [Install packages through 'packageList'](#install-packages-through-packagelist)
        * [Install a package from a third-party repo](#install-a-package-from-a-third-party-repo)
        * [Install a package from SUSE's internal repositories](#install-a-package-from-suses-internal-repositories)
    * [Side-load RPMs](#side-load-rpms)
        * [RPM with dependency resolution from a third-party repository](#rpm-with-dependency-resolution-from-a-third-party-repository)
        * [RPM with depdendency resolution from SUSE's internal repositories](#rpm-with-depdendency-resolution-from-suses-internal-repositories)
    * [Installing unsigned packages](#installing-unsigned-packages)
3. [Package installation workflow](#package-installation-workflow)
    * [Running the EIB container](#running-the-eib-container)
    * [Building the EIB image](#building-the-eib-image)
        * [RPM resolution process](#rpm-resolution-process)    
        * [Troubleshooting](#troubleshooting)
    * [Booting the EIB image](#booting-the-eib-image)

## Supported systems
EIB's **RPM resolution** process and package installation has been tested on the following `x86_64` systems: 

> 1. [Ubuntu 22.04](https://releases.ubuntu.com/jammy/)

Reviewer: Same nit about prioritizing our products.

2. [SLES 15-SP5](https://www.suse.com/download/sles/)
3. [openSUSE Tumbleweed](https://get.opensuse.org/tumbleweed/)
4. [Fedora Linux](https://fedoraproject.org/server/download)

> ## Configure packages for installation

Reviewer: I don't think "configure" is the right word to use here. You're simply "specifying" which packages to install. To me, "configure" means we're providing specific information to manipulate a package on a per package basis.

You can configure packages for installation in the following ways:
1. provide a `packageList` configuration under `operatingSystem.packages` in the EIB image configuration file

> 2. create a `rpms` directory under EIB's configuration directory and provide local RPM files that you want to be installed on the image

Reviewer: Same here about the markdown syntax for ordered lists: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#lists

### Install packages through 'packageList'
To install a package using the `packageList` configuration, at a minimum you must configure the following under `operatingSystem.packages`:

> 1. valid package names under `packageList`

Reviewer: It has just dawned on me, but what if they want to specify a version of the package? Can the user indicate something like `git>=2.3` or does it have to be the specific package name?

Author: We do not have logic for specifying versioning for a package like `>=x.y`. 

We support the following use-case... If there are more than one version of a specific package in a RPM repository (say `git` has `2.35.3-150300.10.33.1`, `2.35.3-150300.10.30.1` and `2.35.3-150300.10.27.1` versions), then the user could specify the desired package version by doing the following config:

```yaml
...
operatingSystem:
  packages:
    packageList:
      - git-2.35.3-150300.10.27.1
```
If for some reason this is not enough, the user could always download the RPM that the he wants to use and put it under the `rpms` directory. 

If you think that this is not enough we could always open a follow-up issue for supporting `foo>=x.y` logic.

> 2. `additionalRepos` or `sccRegistrationCode`

Reviewer: Here too

Below you can find examples for the aforementioned use cases.

#### Install a package from a third-party repo
```yaml
operatingSystem:
  packages:
    packageList:
      - reiserfs-kmp-default-debuginfo
    additionalRepos:
      - url: https://download.opensuse.org/repositories/Kernel:/SLE15-SP5/pool
```
> **_NOTE:_** Before adding any repositories under `additionalRepos`, make sure that they are signed with a valid GPG key. **All non-signed additional repositories will cause EIB to fail.**

#### Install a package from SUSE's internal repositories
```yaml
operatingSystem:
  packages:
    packageList:
      - wget2
    sccRegistrationCode: <your-reg-code>
```

### Side-load RPMs

> Sometimes you may want to install RPM files that are local to your machine. For this use-case, you should create a directory called `rpms` under `<eib-config-dir>/rpms` and copy your local RPM files there.

Reviewer: Nit: Consider rephrasing this as "you may want to install RPM files that are not hosted in a repository." That's a bit broader of a rationale than simply having the file on your machine.

> **_NOTE:_** You must provide an `additionalRepos` entry or a `sccRegistrationCode` if your RPMs are dependent on other packages.

All RPMs that will be side-loaded must have **valid** GPG signatures. The GPG keys used to sign the RPMs must be copied to a directory called `gpg-keys` which must be created under `<eib-config-dir>/rpms`. **Trying to install RPMs that are unsgined or have unrecognized GPG keys will result in a failure of the EIB build process.**

#### RPM with dependency resolution from a third-party repository  
EIB configuration directory tree:

> ```shell

Reviewer: This is great, it's really useful that you provided a model of what it should look like.

```

#### RPM with depdendency resolution from SUSE's internal repositories
EIB configuration directory tree:
```shell
.
├── eib-config-iso.yaml
├── images
│   └── SLE-Micro.x86_64-5.5.0-Default-RT-GM.raw
└── rpms
    ├── gpg-keys
    │   └── git.key
    └── git-2.35.3-150300.10.33.1.x86_64.rpm
```

EIB config file `packages` configuration:
```yaml
operatingSystem:
  packages:
    additionalRepos: <your-reg-code>
```

### Installing unsigned packages
By default EIB does GPG validation for every **additional repository** and **side-loaded RPM**. If you wish to use unsigned additional repositories and/or unsinged RPMs you must add the `noGPGCheck: true` property to EIB's `packages` configuration, like so:
```yaml
operatingSystem:
  packages:
    noGPGCheck: true
```
By providing this configuration, **all** GPG validation will be **disabled**, allowing you to use non-signed pacakges.

> **_NOTE:_** This property is intended for development purposes only. For production use-cases we encourage users to always use EIB's GPG validation.

## Package installation workflow
The package installation workflow can be separated in three logical parts:

> 1. *Running the EIB container* - how to run the EIB container so that the **RPM resolution** has the needed permissions

Reviewer: Same about the markdown ordered lists.

An example of the command can be seen below:
```shell
podman run --rm --privileged -it \
-v $IMAGE_DIR:/eib eib:dev /bin/eib \
-config-file $CONFIG_FILE.yaml \
-config-dir /eib \
-build-dir /eib/_build
```

> **_NOTE:_** Depending on the `cgroupVersion` that Podman operates with, you might also need to run the command with `root` permissions. This is the case for `cgroupVersion: v1`. In this version, non-root usage of the `--privileged` option is not supported. For `cgroupVersion: v2`, non-root usage is supported. 
>
>In order to check the `cgroupVersion` that Podman operates with, run the following command:
>```shell
>podman info | grep cgroupVersion
>```

Once the EIB container has been successfully executed, it parses all the user provided configuration and begins the **RPM resolution** process. 

> ### Building the EIB image

Reviewer: This feels like it should be its own doc covering the design/implementation. The top half of this is much more user based. We don't really have a huge precedent for this yet, but could you split this part out into a subdirectory under docs called `design` so we have an example in place for when (read: if) we get to documenting things more thoroughly? I imagine the k8s and user workload work also needs some sort of design doc in there.

During this phase, EIB prepares the user configured packages for installation. This process is called **RPM resolution** and it includes:

> 1. Validating that each provided package has a GPG signature or comes from a GPG signed RPM repository

Reviewer: Markdown list

2. Resolving and downloading the dependencies for each configured package
3. Creating a RPM repository that consists of the configured packages and their dependencies
4. Configure the usage of this repositry for package installation during the **combustion** phase of the EIB image boot

Below you can find a more detailed description of EIB's **RPM resolution** process. 

> #### RPM resolution process

Reviewer: Also for the design doc

# Installing packages
This documentation dives deeper into how a user can configure packages for installation inside the EIB image. Furthermore, it explains the **RPM resolution** process that EIB goes through so it can ensure that configured packages will be successfully installed even in an **air-gapped** environment.

## Supported systems
EIB's **RPM resolution** process and package installation has been tested on the following `x86_64` systems: 
1. [SLES 15-SP5](https://www.suse.com/download/sles/)
1. [openSUSE Tumbleweed](https://get.opensuse.org/tumbleweed/)
1. [Ubuntu 22.04](https://releases.ubuntu.com/jammy/)
1. [Fedora Linux](https://fedoraproject.org/server/download)

## Specify packages for installation
You can configure packages for installation in the following ways:

> 1. provide a `packageList` configuration under `operatingSystem.packages` in the EIB image configuration file
> 1. create a `rpms` directory under EIB's configuration directory and provide local RPM files that you want to be installed on the image

Reviewer 2: ```suggestion
1. Providing a `packageList` configuration under `operatingSystem.packages` in the EIB image configuration file
1. Creating a `rpms` directory under EIB's configuration directory and providing local RPM files that you want to be installed on the image
```

### Install packages through 'packageList'
To install a package using the `packageList` configuration, at a minimum you must configure the following under `operatingSystem.packages`:

> 1. valid package names under `packageList`
> 1. `additionalRepos` or `sccRegistrationCode`

Reviewer 2: Nit: This is user facing documentation mostly so more details can be helpful. Ignore if you prefer it as it currently is.
```suggestion
1. Valid package names under `packageList`
1. Valid RPM repositories under `additionalRepos` or an SCC registration code under `sccRegistrationCode`
```

> **_NOTE:_** You must provide an `additionalRepos` entry or a `sccRegistrationCode` if your RPMs are dependent on other packages.

All RPMs that will be side-loaded must have **valid** GPG signatures. The GPG keys used to sign the RPMs must be copied to a directory called `gpg-keys` which must be created under `<eib-config-dir>/rpms`. **Trying to install RPMs that are unsgined or have unrecognized GPG keys will result in a failure of the EIB build process.**

#### RPM with dependency resolution from a third-party repository  
EIB configuration directory tree:
```shell
.
├── eib-config-iso.yaml
├── base-images
│   └── SLE-Micro.x86_64-5.5.0-Default-RT-GM.raw
└── rpms
    ├── gpg-keys
    │   └── reiserfs-kpm-default-debuginfo.key
    └── reiserfs-kmp-default-debuginfo-5.14.21-150500.205.1.g8725a95.x86_64.rpm
```

EIB config file `packages` configuration:
```yaml
operatingSystem:
  packages:
    sccRegistrationCode:
      - url: https://download.opensuse.org/repositories/Kernel:/SLE15-SP5/pool
```

> #### RPM with depdendency resolution from SUSE's internal repositories

Reviewer 2: ```suggestion
#### RPM with dependency resolution from SUSE's internal repositories
```

EIB configuration directory tree:
```shell
.
├── eib-config-iso.yaml
├── base-images
│   └── SLE-Micro.x86_64-5.5.0-Default-RT-GM.raw
└── rpms
    ├── gpg-keys
    │   └── git.key
    └── git-2.35.3-150300.10.33.1.x86_64.rpm
```

EIB config file `packages` configuration:
```yaml
operatingSystem:
  packages:
    additionalRepos: <your-reg-code>
```

### Installing unsigned packages

> By default EIB does GPG validation for every **additional repository** and **side-loaded RPM**. If you wish to use unsigned additional repositories and/or unsinged RPMs you must add the `noGPGCheck: true` property to EIB's `packages` configuration, like so:

Reviewer 2: ```suggestion
By default EIB does GPG validation for every **additional repository** and **side-loaded RPM**. If you wish to use unsigned additional repositories and/or unsigned RPMs you must add the `noGPGCheck: true` property to EIB's `packages` configuration, like so:
```

```yaml
operatingSystem:
  packages:
    noGPGCheck: true
```

> By providing this configuration, **all** GPG validation will be **disabled**, allowing you to use non-signed pacakges.

Reviewer 2: ```suggestion
By providing this configuration, **all** GPG validation will be **disabled**, allowing you to use non-signed packages.
```
